Wire Harness: Oxygen Sensor

inceptionev opened the pull request:
This PR creates the design and documentation for the oxygen sensor harness.

Closes #1283 
Updates #1260 

## Self-review checklist:

- [x] Self-review: looked through the `Files changed` tab, browsed repository in branch
- [x] Documentation updated - reflects changes in code, electrical or mechanical design
- [x] Documentation and graphics follow the [documentation style guide](https://github.com/RespiraWorks/Ventilator/wiki/Documentation-style-guide)
- [x] New content is linked, easily discoverable, does not require too many clicks
- [x] Follows other relevant parts of [contributor wiki](https://github.com/RespiraWorks/Ventilator/wiki)
- [x] PR has a descriptive name
- [x] Tagged relevant reviewers

martukas commented:
Scripts need to be modified to generate the newly added docs. I will butt into this branch and add those soon. Please don't merge yet :)

martukas commented:
Ok, scripts have been updated and everything should be much more automated now.
* please note that it was actually documented in `docs/README.md` how to use the docs generation. This document has been updated.
* json was not correctly formatted. I fixed it enough so that it builds. There are, I think, further improvements it could use, but i will save that for the review
* please note failing pre-commit checks
* new document is not linked from main hierarchy, therefore is not visible and likely is not deploying at all because of this, so will not be seen in github pages until that is done
* deployment script temporarily has branch check disabled so it deploys from this branch. Please **DO NOT** merge until that condition has been removed.

martukas commented on docs/wiring/oxygen_sensor.yml:
```diff
@@ -0,0 +1,55 @@
+  - 
+    - X1: [1,3]
+    - W1: [1,2]
+    - X2: [1,3]
```
pre-commit checks should fix this end-of-file problem

martukas commented on docs/purchasing/parts.json:
```diff
@@ -134,5 +134,90 @@
+    ]
+  },
+  {
+    "pn": "ew-26awg-kit",
```
should we maybe just list a specific wire here rather than kit? This might be the more practical thing to buy for us right now, but will skew pricing estimates and will not be good for when ordering with contract manufacturers in the future

martukas replied:
Another idea is to list the specific wires needed and provide nominal price based on that, BUT then add an additional purchasing source with a link to a kit? Yeah it's technically not the same product, but we can maybe add a brief warning next to the supplier name or something.

martukas commented on docs/purchasing/parts.json:
```diff
@@ -134,5 +134,90 @@
+    ]
+  },
+  {
+    "pn": "ec-microfit-f-crimp-20-24awg",
```
I would maybe keep our internal part numbers shorter and less esoteric. When naming imported components in the CAD directories, my rule of thumb is:

Are we likely to ever have 2 parts that are alike in every way other than this one diameter measurement or feature detail? 
* if yes, then it's ok to append it to file name / part descriptor
* if no, then it serves no purpose

So detailed enough to get the gist of what it is and differentiate, not to encode as much info as possible.

inceptionev replied:
Yeah, we will definitely have male (m) and female (f) crimps and very likely also multiple sizes of these.  'crimp' probably also needs to be defined because both housings and connectors go under the ec- prefix.  If we split ec- (connectors) and ep- (pins) we could probably drop -crimp-.

Also, there are so many different options when ordering these contacts, and the wire gauge and gender are the ones I've seen most frequently mixed up by purchasing (not here, but at previous companies), so I put them in the pn.

martukas commented on docs/purchasing/parts.json:
```diff
@@ -134,5 +134,90 @@
+    "manufacturer": "Remington",
+    "mpn": "26UL1007STRKIT10COLOR",
+    "avg_price": 42.35,
+    "price_unit": "ft",
```
I think I may have partially fixed this. This field actually names the units. Would prefer metric here as likely the documentation will use that too.
The "price_units" field in a "supplier" listing below are then a denominator in these units listed here for how big of a batch is being sold.
I see how this could be cause for confusion and I'm soliciting ideas for how to name these more intuitively.

martukas replied:
Ok, so this is a bit nitpickty, but I recommend "price_unit" to be "m", which would change the "price_units" below to be converted, and then "avg_price" above would be divided out to be the per meter price. At the very least it is definitely not the case that price per foot is 106.64USD as it appears now.
See definition for "ew-shrink5" above as example.

Again, I concede that "price_unit" and "price_units" is confusing, so if you have suggestions for how we should name these to be more intuitive, I will make a PR to change the convention.

martukas reviewed (changes requested):
In addition to my comments on specific lines, the main thing that's missing is to add this assembly to the table of contents in the wiring section, so that it actually be browsed to.

inceptionev commented:
for things like wire, it might be a good idea to have a generic RW p/n like #ew-silicone-24awg-green and then have a list of sources for that type of wire.

We should definitely call out the wire gauge and the insulation as these make a material difference, but I am on the fence on whether to include the color in the part number.  Thoughts?

inceptionev commented:
- Added link to the harness table of contents rst
- Corrected 24awg wire kit part numbers.
- I recommend keeping this as a wire kit for now, and deferring the selection of individual wire part numbers to a future PR.

martukas reviewed (changes requested):
The one nitpicky change for wiring price, and fix the problems caught by pre-commit checks.

inceptionev commented:
> The one nitpicky change for wiring price, and fix the problems caught by pre-commit checks.

Thanks for catching that, the pricing should be fixed now.  Still need to look over the pre-commit checks.

inceptionev commented:
it seems to be failing on the 'trim trailing whitespace' pre-commit hook, but it's not clear to me why this is failing.  Any help here would be appreciated.

martukas reviewed (approved):
Looks good now.
